Python/Product/TestAdapter/PythonRunSettings.cs

```diff
@@ -0,0 +1,227 @@
+        private readonly IComponentModel _compModel;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly Dispatcher _dispatcher;
+        internal static Uri PythonCodeCoverageUri = new Uri("datacollector://Microsoft/PythonCodeCoverage/1.0");
+        internal static string CodeCoverageUriString = @"datacollector://Microsoft/CodeCoverage/2.0";
+        internal const int cmdidImportCoverage = 0x10f;
+        [ImportingConstructor]
+        public PythonRunSettings([Import(typeof(SVsServiceProvider))]IServiceProvider serviceProvider) {
+            _compModel = (IComponentModel)serviceProvider.GetService(typeof(SComponentModel));
+            var opState = _compModel.GetService<IOperationState>();
+            opState.StateChanged += StateChange;
+            _dispatcher = Dispatcher.CurrentDispatcher;
+            _serviceProvider = serviceProvider;
+        }
+
+        private void StateChange(object sender, OperationStateChangedEventArgs e) {
+            if (e.State == TestOperationStates.TestExecutionFinished) {
+                var resultUris = e.Operation.GetRunSettingsDataCollectorResultUri(PythonCodeCoverageUri);
+                if (resultUris != null) {
+                    foreach (var eachAttachment in resultUris) {
+                        string filePath = eachAttachment.LocalPath;
+
+                        if (!string.IsNullOrEmpty(filePath) && File.Exists(filePath)) {
+                            object inObj = filePath;
+                            object outObj = null;
+
+                            var dte = (EnvDTE.DTE)_serviceProvider.GetService(typeof(EnvDTE.DTE));
+                            dte.Commands.Raise(
+                                GuidList.guidPythonToolsCmdSet.ToString("B"),
+                                cmdidImportCoverage,
+                                ref inObj,
+                                ref outObj
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        public ICodeCoverageSettingsService CodeCoverage {
+            get {
+                return _compModel.GetService<ICodeCoverageSettingsService>();
+            }
+        }
+
+        public string Name {
+            get {
+                return "Python Run Settings";
+            }
+        }
+
+        public IXPathNavigable AddRunSettings(IXPathNavigable inputRunSettingDocument, IRunSettingsConfigurationInfo configurationInfo, ILogger log) {
+            XPathNavigator navigator = inputRunSettingDocument.CreateNavigator();
+            var python = navigator.Select("/RunSettings");
+            if (python.MoveNext()) {
+                using (var writer = python.Current.AppendChild()) {
+                    var pyContainers = configurationInfo.TestContainers
+                        .Where(x => x is TestContainer)
+                        .Cast<TestContainer>()
+                        .GroupBy(x => x.Project);
```

Comment:
Instead of .Where().Cast<T>()

---
In reply to: [68104358](https://github.com/Microsoft/PTVS/pull/1337#discussion_r68104358) [](ancestors = 68104358)